research-software-directory/RSD-as-a-service
# 705 tos agreement

# Require users to agree to Terms and Data Privacy

Fixes #705

Changes proposed in this pull reques:

* require users to agree to Terms of Service and state that they have read the data privacy declaration
* users cannot perform any action until they agree to all the documents
* if the user did not yet agree, a modal will pop up wherever something could be added or edited
* renamed 
* the user agreement can additionally be configured in the user profile
* a "Profile settings" was added to the user menu

How to test:

* clean volumes, rebuild database/frontend and start
  ```
  docker-compose down --volumes
  docker-compose build database frontend
  docker-compose up --scale scrapers=0
  ```
* login as a new user, e.g. as a local user
* try to add a new software, this modal will popup 
  ![modal](https://user-images.githubusercontent.com/14222414/215996740-b3623889-53cb-445e-82e8-6e6965035086.png)
* click on Cancel -> you will be returned to page you came from
* try to add software again
* agree to the Terms of Service and the Privacy Statement, the OK button will turn green, click ok
* add a new software
* open the user Menu -> Profile Settings
* the switches below "User agreements" should be activated
* deactivate the switches
* Go to "My Software" and try to edit the software -> modal shows
* the same happens when trying to edit organisations where the user is maintainer of or when creating/editing projects

Things that should/could be improved

~~* the check whether a user agreed to all the documents is performed in the database. Maybe the NLESC doesn't require the agreement to three documents (e.g. if you will only have Terms of Service), so a configuration about which documents are required could be useful. Since checks are done in the DB, it might be worth to have some configuration options in the DB as well? (@ewan-escience ?)~~
~~* If the user logs in the first time, they should be forwarded directly to a separate page where they will be asked to document their agreement~~

To do

* [x] add check for DELETE
* [x] Rename Data Protection Declaration to Privacy Statement

Many thanks to @dmijatovic for fixing the issue with the forms :)

PR Checklist:
